sql/core/src/main/scala/org/apache/spark/sql/artifact/ArtifactManager.scala

```diff
@@ -68,18 +69,66 @@ class ArtifactManager(session: SparkSession) extends Logging {
       s"$artifactRootURI${File.separator}${session.sessionUUID}")
 
   // The base directory/URI where all class file artifacts are stored for this `sessionUUID`.
-  protected[artifact] val (classDir, classURI): (Path, String) =
+  protected[artifact] val (classDir, replClassURI): (Path, String) =
     (ArtifactUtils.concatenatePaths(artifactPath, "classes"),
       s"$artifactURI${File.separator}classes${File.separator}")
 
-  protected[artifact] val state: JobArtifactState =
-    JobArtifactState(session.sessionUUID, Option(classURI))
 
-  def withResources[T](f: => T): T = {
-    Utils.withContextClassLoader(classloader) {
-      JobArtifactSet.withActiveJobArtifactState(state) {
+  private lazy val sessionIsolated =
+    session.conf.get("spark.session.isolate.artifacts", "true") == "true"
+  private lazy val replIsolated =
+    session.conf.get("spark.repl.isolate.artifacts", "false") == "true"
+
+  protected[sql] lazy val state: JobArtifactState = {
+    (sessionIsolated, replIsolated) match {
+      case (true, true) => JobArtifactState(session.sessionUUID, Some(replClassURI))
+      case (true, false) => JobArtifactState(session.sessionUUID, None)
+      case (false, true) => throw SparkException.internalError(
+        "To enable REPL isolation, session isolation must also be enabled.")
+      case (false, false) => null
+    }
+  }
+
+  private val shouldApplyClassLoader = new AtomicBoolean(false)
```

Comment:
Can we add a short blurb on when this boolean should be set to true?

Reply:
Done!